### PR TITLE
Raptor: add a crow fly when no street network

### PR DESCRIPTION
**Pull request by pbougue**
add a crow fly when no street network at the departure or arrival
fix http://jira.canaltp.fr/browse/NAVP-201

Use-case :
For fr-ne, Gare de l'Est (which is not in the street network coverage) to Amiens : http://api.navitia.io/v1/coverage/fr-ne/journeys?from=stop_area:SIN:SA:OCE87113001&to=stop_area:SIN:SA:OCE87313874
Before there was no direct path from Gare de l'Est to Gare du Nord (which is also not in the coverage, but more interesting for this journey), now a crow-fly is created between the two stop areas.

### CODE REVIEW

**Comment by buildbotctp** [comment]
Can one of the admins verify this patch?

**Comment by TeXitoi** [comment]
To-do: check documentation.

**Comment by pbougue** [comment]
For demonstration, on fr-sncfnational, a journey from Gare de l'Est to Lille Europe, uses section from Gare du Nord to gare de Lille Flandres : http://localhost:5000/v1/coverage/navitia/journeys?from=stop_area:RAT:SA:grest&to=stop_area:OCE:SA:87223263

File: source/georef/street_network.cpp [hunk 1]
```diff
@@ -237,16 +237,27 @@ void PathFinder::start_distance_dijkstra(navitia::time_duration radius) {
...
 
 std::vector<std::pair<type::idx_t, navitia::time_duration>>
+PathFinder::crow_fly_find_nearest_stop_points(navitia::time_duration radius, const proximitylist::ProximityList<type::idx_t>& pl) {
+    // Searching for all the elements that are less than radius meters awyway in crow fly
+    float crow_fly_dist = radius.total_seconds() * speed_factor * georef::default_speed[mode];
+
+    std::vector< std::pair<nt::idx_t, type::GeographicalCoord> > elements = pl.find_within(start_coord, crow_fly_dist);
+
+    std::vector<std::pair<type::idx_t, navitia::time_duration>> result;
+    for (auto element: elements) {
```
Comment (antoine-de) [inline source/georef/street_network.cpp]: not very important there, but remember, that this is doing a copy (of 3 int, so here it's not mandatory)

File: source/routing/raptor_api.cpp [hunk 2]
```diff
@@ -305,48 +305,53 @@ static void add_pathes(EnhancedResponse& enhanced_response,
...
-                if(temp.path_items.size() > 0) {
-                    //because of projection problem, the walking path might not join
-                    //exactly the routing one
-                    nt::GeographicalCoord routing_first_coord = departure_stop_point->coord;
-                    if (temp.path_items.back().coordinates.back() != routing_first_coord) {
-                        //if it's the case, we artificialy add the missing segment
-                        temp.path_items.back().coordinates.push_back(routing_first_coord);
-                    }
+            }
+            else if(!sn_departure_path.path_items.empty()) {
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: Please, space between `if` and `(`

Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: better `} else if (`

File: source/routing/raptor_api.cpp [hunk 3]
```diff
@@ -305,48 +305,53 @@ static void add_pathes(EnhancedResponse& enhanced_response,
...
+            // no crow fly section, but we'll have to update the start of the journey
             //first is zonal ODT, we do nothing, there is no walking
         }
-        else if (!path.items.front().stop_points.empty()
-                && use_crow_fly(origin, path.items.front().stop_points.front(), d)){
-            const auto sp_dest = path.items.front().stop_points.front();
-            type::EntryPoint destination_tmp(type::Type_e::StopPoint, sp_dest->uri);
-            bt::time_period action_period(path.items.front().departures.front(),
-                                          path.items.front().departures.front()+bt::minutes(1));
+        else if (!path.items.empty() && !path.items.front().stop_points.empty()){
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: better `} else if (!path.items.empty() && !path.items.front().stop_points.empty()) {` (with a space between `)` and `{`, and `}` and `else` on the same line)

File: source/routing/raptor_api.cpp [hunk 4]
```diff
@@ -305,48 +305,53 @@ static void add_pathes(EnhancedResponse& enhanced_response,
...
             const auto& departure_stop_point = path.items.front().stop_points.front();
-            auto temp = worker.get_path(departure_stop_point->idx);
-            fill_crowfly_section(origin, destination_tmp, get_crowfly_mode(temp), path.items.front().departures.front(),
-                                 d, enhanced_response, pb_journey, now, action_period);
+            georef::Path sn_departure_path = worker.get_path(departure_stop_point->idx);
+
+            if (use_crow_fly(origin, path.items.front().stop_points.front(), sn_departure_path, d)){
+                const auto sp_dest = path.items.front().stop_points.front();
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: `const auto&` is a good habbit (but here, we don't care).

File: source/routing/raptor_api.cpp [hunk 5]
```diff
@@ -305,48 +305,53 @@ static void add_pathes(EnhancedResponse& enhanced_response,
...
+                type::EntryPoint destination_tmp(type::Type_e::StopPoint, sp_dest->uri);
+                bt::time_period action_period(path.items.front().departures.front(),
+                                              path.items.front().departures.front()+bt::minutes(1));
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: spaces around `+`

File: source/routing/raptor_api.cpp [hunk 6]
```diff
@@ -305,48 +305,53 @@ static void add_pathes(EnhancedResponse& enhanced_response,
...
+                const time_duration& crow_fly_duration = find_or_default(
+                        SpIdx(*path.items.front().stop_points.front()),
+                        worker.departure_path_finder.distance_to_entry_point);
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: :+1:

File: source/routing/raptor_api.cpp [hunk 7]
```diff
@@ -305,48 +305,53 @@ static void add_pathes(EnhancedResponse& enhanced_response,
...
-                        fill_pb_placemark(origin, d, first_section->mutable_origin(), 2, now, action_period, show_codes);
-                        // We add coherence with the first pt section
-                        last_section->mutable_destination()->Clear();
-                        fill_pb_placemark(departure_stop_point, d, last_section->mutable_destination(), 2, now, action_period, show_codes);
-                    }
+                const auto walking_time = sn_departure_path.duration;
+                departure_time = path.items.front().departure - walking_time.to_posix();
+                fill_street_sections(enhanced_response, origin, sn_departure_path, d, pb_journey,
+                    departure_time);
+                if(pb_journey->sections_size() > 0){
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: spacing

File: source/routing/raptor_api.cpp [hunk 8]
```diff
@@ -517,51 +522,55 @@ static void add_pathes(EnhancedResponse& enhanced_response,
             // last is 'taxi like' ODT, we do nothing, there is no walking nor crow fly section,
             // but we have to updated the end of the journey
         }
-        else if (!path.items.back().stop_points.empty()
-                && use_crow_fly(destination, path.items.back().stop_points.back(), d)){
-            const auto sp_orig = path.items.back().stop_points.back();
-            type::EntryPoint origin_tmp(type::Type_e::StopPoint, sp_orig->uri);
-            bt::time_period action_period(path.items.back().departures.back(),
-                                          path.items.back().departures.back()+bt::minutes(1));
+        else if (!path.items.empty() && !path.items.back().stop_points.empty()) {
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: spacing

File: source/routing/routing.cpp [hunk 9]
```diff
@@ -73,18 +73,28 @@ std::string PathItem::print() const {
 }
 
 
-bool use_crow_fly(const type::EntryPoint& point, const type::StopPoint* stop_point, const type::Data& data){
+bool use_crow_fly(const type::EntryPoint& point, const type::StopPoint* stop_point,
```
Comment (antoine-de) [inline source/routing/routing.cpp]: not sure, but maybe it would be nice to add a block comment to explain the whole mechanism

Comment (antoine-de) [inline source/routing/routing.cpp]: my bad, it's in the header :astonished:

**Comment by kinnou02** [comment]
Do you have check jormungandr filters? some of them use the fallback duuration, we need to be sure that crowfly are included.

**Comment by pbougue** [comment]
Jenkins, test this please.

**Comment by pbougue** [comment]
@kinnou02 : good point, it was already managed, except for geojson output so added a commit for that

File: source/routing/raptor_api.cpp [hunk 10]
```diff
@@ -305,48 +305,51 @@ static void add_pathes(EnhancedResponse& enhanced_response,
...
+                                     get_crowfly_mode(sn_departure_path),
+                                     departure_time, d, enhanced_response,
+                                     pb_journey, now, action_period);
+
+            } else if (!sn_departure_path.path_items.empty()) {
+                //because of projection problem, the walking path might not join
+                //exactly the routing one
+                nt::GeographicalCoord routing_first_coord = departure_stop_point->coord;
+                if (sn_departure_path.path_items.back().coordinates.back() != routing_first_coord) {
+                    //if it's the case, we artificialy add the missing segment
```
Comment (TeXitoi) [inline source/routing/raptor_api.cpp]: artificialLy

File: source/jormungandr/tests/check_utils.py [hunk 11]
```diff
@@ -462,6 +462,7 @@ def is_valid_journey(journey, tester, query):
     request = get_valid_datetime(journey['requested_date_time'])
 
     assert arrival >= departure
+    assert (arrival - departure).seconds - journey['duration'] <= len(journey['sections']) - 1
```
Comment (TeXitoi) [inline source/jormungandr/tests/check_utils.py:463]: please comment this assert

File: source/type/pb_converter.cpp [hunk 12]
```diff
@@ -1071,18 +1071,28 @@ void fill_pb_placemark(const type::EntryPoint& point, const type::Data &data,
...
     section->set_id(response.register_section());
     fill_pb_placemark(origin, data, section->mutable_origin(), 2, now, action_period);
     fill_pb_placemark(destination, data, section->mutable_destination(), 2, now, action_period);
-    section->set_begin_date_time(navitia::to_posix_timestamp(time));
-    section->set_duration(0);
-    section->set_length(0);
-    section->set_end_date_time(navitia::to_posix_timestamp(time));
+    section->set_begin_date_time(navitia::to_posix_timestamp(origin_time));
+    section->set_duration(crow_fly_duration.total_seconds());
+    if(crow_fly_duration.total_seconds() > 0){
```
Comment (TeXitoi) [inline source/type/pb_converter.cpp]: spacing

File: source/type/pb_converter.cpp [hunk 13]
```diff
@@ -1071,18 +1071,28 @@ void fill_pb_placemark(const type::EntryPoint& point, const type::Data &data,
...
+        section->set_length(origin.coordinates.distance_to(destination.coordinates));
+        auto* new_coord = section->add_shape();
+        new_coord->set_lon(origin.coordinates.lon());
+        new_coord->set_lat(origin.coordinates.lat());
+        new_coord = section->add_shape();
+        new_coord->set_lon(destination.coordinates.lon());
+        new_coord->set_lat(destination.coordinates.lat());
+    }else{
```
Comment (TeXitoi) [inline source/type/pb_converter.cpp]: spacing

File: source/type/pb_converter.h [hunk 14]
```diff
@@ -151,8 +151,10 @@ inline pbnavitia::StopArea* get_sub_object(const type::StopArea*, pbnavitia::PtO
 inline pbnavitia::StopPoint* get_sub_object(const type::StopPoint*, pbnavitia::PtObject* pt_object) { return pt_object->mutable_stop_point(); }
 
 
-void fill_crowfly_section(const type::EntryPoint& origin, const type::EntryPoint& destination, type::Mode_e mode, boost::posix_time::ptime time,
-                          const type::Data& data, EnhancedResponse &response,  pbnavitia::Journey* pb_journey,
+void fill_crowfly_section(const type::EntryPoint& origin, const type::EntryPoint& destination,
+                          const time_duration& crow_fly_duration, type::Mode_e mode,
+                          boost::posix_time::ptime origin_time, const type::Data& data,
+                          EnhancedResponse &response,  pbnavitia::Journey* pb_journey,
```
Comment (TeXitoi) [inline source/type/pb_converter.h]: `&` spacing incoherence.

**Comment by TeXitoi** [comment]
OK for me

**Pull request merged**